zcncore/transaction_query_base.go

```diff
@@ -143,52 +144,54 @@ func (tq *TransactionQuery) checkHealth(ctx context.Context, host string) error
 		tq.Unlock()
 
 		if len(tq.offline) >= tq.max {
-			return ErrNoOnlineSharders
+			errCh <- ErrNoOnlineSharders
+			return
 		}
 	}
-
-	return nil
+	errCh <- nil
+	return
 }
 
-// randOne random one health sharder
-func (tq *TransactionQuery) randOne(ctx context.Context) (string, error) {
-
-	randGen := rand.New(rand.NewSource(time.Now().UnixNano()))
-	for {
-
-		// reset selected if all sharders were selected
-		if len(tq.selected) >= tq.max {
-			tq.selected = make(map[string]bool)
-		}
-
-		i := randGen.Intn(len(tq.sharders))
-		host := tq.sharders[i]
-
-		_, ok := tq.selected[host]
-
-		// it was selected, try next
-		if ok {
-			continue
-		}
-
-		tq.selected[host] = true
-
-		err := tq.checkHealth(ctx, host)
-
-		if err != nil {
-			if errors.Is(err, ErrNoOnlineSharders) {
-				return "", err
+// getRandomSharder returns a random healthy sharder
+func (tq *TransactionQuery) getRandomSharder(ctx context.Context) (string, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	finalErrCh := make(chan error)
+	onlineSharderCh := make(chan string)
+	for _, sharder := range tq.sharders {
+		go func(sharder string) {
+			errCh := make(chan error)
+			go tq.checkSharderHealth(ctx, sharder, errCh)
+			select {
+			case <-ctx.Done():
+				finalErrCh <- ctx.Err()
+				return
+			case e := <-errCh:
+				if errors.Is(e, ErrSharderOffline) {
+					tq.RLock()
+					defer tq.RUnlock()
+					if len(tq.offline) >= tq.max {
+						finalErrCh <- ErrNoOnlineSharders
```

Comment:
finalErrCh should be buffered to read all the errors, to unlock the goroutines, or you should force wait for all to finish with wg

Reply:
Made some changes here too, removed the extra goroutine and waiting on ctx.Done, cuz it was unnecessary. Pls have a look now.